### PR TITLE
MEN-7526: Unit Test framework follow-ups

### PR DESCRIPTION
**Pull request by lluiscampos**
Clean-ups and follow-ups from #97

### CODE REVIEW

**Review by vpodzime — COMMENTED** [review]
+ ~loner~ longer in the last commit message.

Looks good to me otherwise.

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -43,17 +45,23 @@ test:smoke-build:weak:
...
     - cmake --build build --parallel $(nproc --all)
     - ./build/mender-mcu-client.elf --help
 
 test:unit:
   stage: test
-  image: debian:12
+  image: debian:12-slim
   before_script:
-    - apt update && apt install -yyq g++ cmake git make lcov pkg-config curl
+    - apt update && apt install -yyq
```
Comment (danielskinstad) [inline .gitlab-ci.yml:60]: I don't think it matters much, but isn't `apt-get` preferred for scripts?

File: cmake/asan.cmake [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# @file      asan.cmake
+# @brief     Address Sanitazier options
```
Comment (danielskinstad) [inline cmake/asan.cmake:2]:
```suggestion
# @brief     Address Sanitizer options
```

File: cmake/helper.cmake [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# @file      helper.cmake
+# @brief     Helper for sanitaziers
```
Comment (danielskinstad) [inline cmake/helper.cmake:2]:
```suggestion
# @brief     Helper for sanitizers
```

File: cmake/threadsan.cmake [hunk 4]
```diff
@@ -1,7 +1,7 @@
-# @file      CMakeLists.txt
-# @brief     Mocks CMakeLists file
+# @file      threadsan.cmake
+# @brief     Thread Sanitazier options
```
Comment (danielskinstad) [inline cmake/threadsan.cmake:2]:
```suggestion
# @brief     Thread Sanitizer options
```

File: cmake/undefsan.cmake [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# @file      undefsan.cmake
+# @brief     Undefined Sanitazier options
```
Comment (danielskinstad) [inline cmake/undefsan.cmake:2]:
```suggestion
# @brief     Undefined Sanitizer options
```

**Review by danielskinstad — APPROVED** [review]

File: tests/unit/CMakeLists.txt [hunk 6]
```diff
@@ -36,9 +36,16 @@ include_directories(${CMAKE_CURRENT_BINARY_DIR}/_deps/cjson-src/)
...
-# Include mender-mcu-client library
+# Include mender-mcu-client library, forcing warnings and errors
+set(CONFIG_MENDER_ALL_WARNINGS_AS_ERRORS ON CACHE BOOL "All warnings as errors" FORCE)
 add_subdirectory("${CMAKE_CURRENT_LIST_DIR}/../../target/posix" build)
 
+# And add the same flags for the test targets too
+add_compile_options(-Wall -Wextra -Werror -Wpointer-arith)
+
+# Silence C/C++ compat warning for structs and unions
+add_compile_options(-Wno-extern-c-compat)
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply tests/unit/CMakeLists.txt:34]: Of course. Thank you.

File: target/posix/CMakeLists.txt [hunk 7]
```diff
@@ -74,14 +74,29 @@ if (CONFIG_MENDER_PROVIDES_DEPENDS)
...
+
+# Depencency: Mbed TLS. Fetch the project from upstream
+# This is because the popular Linux distributions (Fedora, Ubuntu) are stuck in 2.x version
+# due to an breaking API change in 3.x. As Zephyr project uses 3.x, we want to avoid poluting
+# the code with many ifdefs for the two major versions
+include(FetchContent)
+FetchContent_Declare(
+  MbedTLS
+  GIT_REPOSITORY https://github.com/Mbed-TLS/mbedtls
+  GIT_TAG        v3.6.0
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply target/posix/CMakeLists.txt:93]: Here I sneaked in my personal preference of aligning it to Zephyr project... :see_no_evil: 

To be honest, I don't have an strong opinion on this. Having a newer one in the same series should still be fully compatible. So, once you commented on it, I'll update it :grin:

File: tests/smoke/CMakeLists.txt [hunk 8]
```diff
@@ -75,10 +72,10 @@ endif()
 
 # Link the mender-mcu-client library
 if(CONFIG_MENDER_PLATFORM_NET_TYPE MATCHES "generic/curl")
-    target_link_libraries(mender-mcu-client curl)
+    target_link_libraries(mender-mcu-client PRIVATE curl)
 endif()
 if(CONFIG_MENDER_PLATFORM_SCHEDULER_TYPE MATCHES "posix")
-    target_link_libraries(mender-mcu-client rt)
+    target_link_libraries(mender-mcu-client PRIVATE rt)
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply tests/smoke/CMakeLists.txt:78]: To be honest this I don't know where is coming from - is it really the library or the smoke tests? If it is the library, should we move it into the library's `CMakeLists.txt` file?

This part (the smoke tests) I didn't want to rework more than necessary. It has an error (from CMake point of view) in that in uses the `mender-mcu-client` target directly, where what it should do is to have its own target and then link to the library (similar as it is done in the unit tests).

(the above is rather a rant than a reply to you - if you confirm that this is part of the library I'll modify it as public :ok_hand: )

File: target/posix/CMakeLists.txt [hunk 9]
```diff
@@ -22,6 +22,7 @@ cmake_minimum_required(VERSION 3.16.3)
 project(mender-mcu-client LANGUAGES C)
 
 # Include helper to define MENDER_MCU_ variables
+include(../../cmake/CMake_posix_defaults.txt)
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply target/posix/CMakeLists.txt:25]: Yes, thanks.

File: .gitlab-ci.yml (hunk 1 above)
Comment (lluiscampos) [reply .gitlab-ci.yml:60]: It is. But we have this everywhere already. I believe folks in Ubuntu won't dare to change this now after many years...

File: cmake/undefsan.cmake (hunk 5 above)
Comment (lluiscampos) [reply cmake/undefsan.cmake:2]: Thanks for all the typos :see_no_evil:

File: tests/smoke/CMakeLists.txt (hunk 8 above)
[the remark this answers is not in the document]
Comment (vpodzime) [reply tests/smoke/CMakeLists.txt:78]: Oh! Wait, this is wrong in a different way then, right? To me it looks like this should be in `target/posix/CMakeLists.txt` as a `PUBLIC` dependency of the library when building for POSIX, right?

[the remark this answers is not in the document]
Comment (lluiscampos) [reply tests/smoke/CMakeLists.txt:78]: Exactly!

I won't fix it in this PR, though. But we can revisit the smoke tests _target_ on a later point to correct it.

**Pull request merged**